### DISCUSSION TITLE
How can I choose which files and directories get copied to the Dagger volume from a given directory?

### DISCUSSION THREAD

**Discussion by hariscodes** [dagger/dagger#8604] [Help] [answered]
Hello!

We're working with a really large Monorepo, and we'd like to be able to run our Dagger module including only two directories off of the root. (I'm testing this in a separate, smaller repo at the moment)

I've attempted to do this two different ways:
### Defining a repo-specific view in my module's dagger.json file to **include** the files I wanted
```json
{
  "name": "package",
  "sdk": "go",
  "source": ".",
  "engineVersion": "v0.13.3",
  "views": [
    {
      "name": "example",
      "patterns": [
        "frontend",
        "backend",
        "deploy"
      ]
    }
  ]
}
```
Then referencing in the command-line call like this: `dagger -m ssh://git@github.com/example/redacted-repo/package@main --progress plain call --root-dir ".":example`

Which appears to correctly pick up that I am referencing a view:
```
12  :     rootDir: ModuleSource.resolveDirectoryFromCaller(path: ".", viewName: "example"): Directory!
```

But then throws up because nothing is actually getting copied over. I see my local Docker volume has no data in it and I get a nil-pointer exception from attempting to interact with the file system:
```
{"time":"2024-09-30T19:14:18.645860292Z","level":"INFO","msg":"discovering app files","basedir":"./deploy/test"}
Stderr:
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x51d5f8]
```

When I exclude the view from the `dagger call`, the module works as expected, but copies the entire repo over.

### Using `WithoutFiles` to **exclude** files I didn't want:
```go
// --exclude is a flag which takes a comma separated list of glob patterns
filteredRootDir := rootDir.WithoutFiles(exclude)

// The container is constructed using the following function, passing the filteredRootDir as the rootDir parameter
func ContainerRootDir(dag *dagger.Client, ref string, creds *dagger.Secret, rootDir *dagger.Directory) *dagger.Container {
	return dag.
		Container().
		From(ref).
		WithSecretVariable(Credentials, creds).
		WithDirectory(".", rootDir)
}
```
This is at least functional, in that it doesn't fail, but it also doesn't actually exclude the files from being copied to the container volume. It simply copies all the files over, then removes the ones in the `--excludes` list.

Based on my reading of a bunch of different issues and pull requests, I'm not sure what the right way to do this is, or if it's even possible. Any help would be greatly appreciated.

**Comment by vikram-dagger** [comment, marked as answer]
I'm not sure if you already tried this but you can use an `ignore` pattern with `defaultPath` for exclusion: https://docs.dagger.io/manuals/developer/functions#directories-and-files

There are some examples and discussion of this approach also in https://discord.com/channels/707636530424053791/1288245581701058560

**Reply by hariscodes** [reply]
This worked, thanks a bunch! Using the `ignore` + `defaultPath` pattern worked for reducing the file upload size:

```go
// Root directory where all modules/bases are accessible from. (Usually just the $GITHUB_WORKSPACE)
// +optional
// +defaultPath="."
// +ignore=["*", "!*.yaml", "!*.yml"]
rootDir *dagger.Directory,
```

I was also able to allow for optimization of files mounted to my dagger-built container using a `--include` and `--exclude` flag with `DirectoryWithDirectoryOpts`:

```go
dirOpts := dagger.DirectoryWithDirectoryOpts{
                 // these are both input parameters to my module
		Include: include,
		Exclude: exclude,
	}
```

**Reply by hariscodes** [reply]
I think there might be a bug of some kind here. I can still see some non-yaml files in some of my snapshots, despite wiping out my entire dagger cache before re-running:
<img width="931" alt="image" src="https://github.com/user-attachments/assets/2625069a-723b-464b-92dd-6da149e16320">

**Reply by helderco** [reply]
There's some tips on how to debug this in:
- https://github.com/dagger/dagger/issues/8594

**Comment by skycaptain** [comment]
Jumping on this discussion as I'm investigating the same issue. I'm trying to implement the pattern mentioned [here](https://docs.dagger.io/cookbook#continue-using-a-container-after-command-execution-fails), where we have a long-running build in CI and want to return multiple test results along with the exit code. With the proposed setup, we'd need to call Dagger multiple times to receive individual fields from the `TestResult` class—e.g first to get the exit code, then subsequent times to get the various artifacts. However, the latter modifies the directory we're passing to Dagger, invalidating the cache and forcing the build to rerun on subsequent calls. That's why I'm exploring using `Ignore` or views to exclude the directory where artifacts are stored. I've encountered some issues, though:

Consider this minimal example:

```python
from typing import Annotated

import dagger
from dagger import DefaultPath, Ignore, function, object_type

@object_type
class Test:
    @function
    def get_dir(
        self, src: Annotated[dagger.Directory, DefaultPath("."), Ignore(["sdk"])]
    ) -> dagger.Directory:
        return src
```

When called without arguments (using the provided default path), it works as expected, excluding `sdk`:

```shell
$ dagger call get-dir entries

.gitattributes
.gitignore
dagger.json
pyproject.toml
src
uv.lock
```

However, when running the same command but providing `--src`, `sdk` is included:

```shell
$ dagger call get-dir --src . entries

.gitattributes
.gitignore
dagger.json
pyproject.toml
sdk
src
uv.lock
```

If I modify the code to remove `DefaultPath`:

```python
from typing import Annotated

import dagger
from dagger import Ignore, function, object_type

@object_type
class Test:
    @function
    def get_dir(self, src: Annotated[dagger.Directory, Ignore(["sdk"])]) -> dagger.Directory:
        return src
```

Interestingly, `Ignore` seems to work only in combination with `DefaultPath`. The same behavior occurs:

```shell
$ dagger call get-dir --src .  entries

.gitattributes
.gitignore
dagger.json
pyproject.toml
sdk
src
uv.lock
```

It appears that `Ignore` only works when the user doesn't provide an argument, i.e., doesn't override the `DefaultPath`. Is this the intended behavior, or am I missing something important here? I've also tried views, which work as expected, but require the user to know which view to use. To me, the documentation doesn't clearly explain when to use which variant.

**Reply by helderco** [reply]
When it was released, the `Ignore` meta only applied to `DefaultPath`. With [v0.13.4](https://github.com/dagger/dagger/releases/tag/v0.13.4) it will filter directories that are explicitly passed too ([PR](https://github.com/dagger/dagger/pull/8436)). Note that it's still not supported when passing a directory to a dependency (via code).

**Reply by skycaptain** [reply]
Thanks, I can confirm that `Ignore` now works as expected. 

> Note that it's still not supported when passing a directory to a dependency (via code).

That's fine. My main use case is to:

a) Limit the amount of data we need to send to the engine, as we have quite large (several GBs) build directories.
b) Address GitLab CI's requirement of placing artefacts under `$CI_PROJECT_DIR`, which invalidates the cache as mentioned above.

With `Ignore` we can address both these issues without requiring users to know about (and likely forget) using views.

**Comment by vikram-dagger** [comment]
@marcosnils @TomChv can you add your thoughts?

**Comment by vikram-dagger** [comment]
> Interestingly, Ignore seems to work only in combination with DefaultPath

@skycaptain I believe this is fixed in https://github.com/dagger/dagger/pull/8436 and available in Dagger 0.13.5.
